File: go/p1.go
package main

import "fmt"

// 利用map来实现  复杂度o(n)
func twoSum(nums []int, target int) []int {
	nMap := make(map[int]int, len(nums))
	for i := 0; i < len(nums); i++ {
		if v, ok := nMap[target-nums[i]]; ok {
			return []int{v, i}
		}
		nMap[nums[i]] = i
	}
	return []int{}
}

func main() {
	nums := []int{2, 7, 11, 15}
	fmt.Println(twoSum(nums, 9))
}

File: python/p1.py
class Solution(object):
    # 利用 map来存每个数出现的下标
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        nMap = {}
        for i, num in enumerate(nums):
            v = nMap.get(target - num, -1)
            if v >= 0:
                return [v, i]
            nMap[num] = i
        return []


if __name__ == "__main__":
    s = Solution()
    print(s.twoSum([1, 2, 3, 4], 3))
